core/services/ocr2/plugins/ocr2keeper/evmregistry/v21/logprovider/buffer_v2.go

```diff
@@ -0,0 +1,324 @@
+	ub.maxLogs.Store(int32(maxLogs))
+}
+
+func (ub *upkeepLogBuffer) size() int {
+	ub.lock.RLock()
+	defer ub.lock.RUnlock()
+
+	return len(ub.q)
+}
+
+func (ub *upkeepLogBuffer) dequeue(start, end int64, limit int) ([]logpoller.Log, int) {
+	ub.lock.Lock()
+	defer ub.lock.Unlock()
+
+	if len(ub.q) == 0 {
+		return nil, 0
+	}
+
+	var results []logpoller.Log
+	var remaining int
+	updatedLogs := make([]logpoller.Log, 0, ub.maxLogs.Load())
+	for _, l := range ub.q {
+		if l.BlockNumber >= start && l.BlockNumber <= end {
+			if len(results) < limit {
+				results = append(results, l)
+				continue
+			}
+			remaining++
+		}
+		updatedLogs = append(updatedLogs, l)
+	}
+
+	if len(results) > 0 {
+		ub.q = updatedLogs
+	}
+
+	prommetrics.AutomationLogsInLogBuffer.Sub(float64(len(results)))
+
+	return results, remaining
+}
+
+func (ub *upkeepLogBuffer) enqueue(offsetBlock int64, logsToAdd ...logpoller.Log) (int, int) {
+	ub.lock.Lock()
+	defer ub.lock.Unlock()
+
+	logs := ub.q
+	var added int
+	for _, log := range logsToAdd {
+		if log.BlockNumber < offsetBlock {
+			ub.lggr.Debugw("Skipping log from old block", "offsetBlock", offsetBlock, "logBlock", log.BlockNumber)
+			continue
+		}
+		logid := logID(log)
+		if _, ok := ub.visited[logid]; ok {
+			ub.lggr.Debugw("Skipping known log", "offsetBlock", offsetBlock, "logBlock", log.BlockNumber)
+			continue
+		}
+		added++
+		if len(logs) == 0 {
+			logs = append(logs, log)
+		} else {
+			i, _ := sort.Find(len(logs), func(i int) int {
+				return LogComparator(log, logs[i])
+			})
+			if i == len(logs) {
+				logs = append(logs, log)
+			} else {
+				logs = append(logs[:i], append([]logpoller.Log{log}, logs[i:]...)...)
+			}
+		}
+		ub.visited[logid] = log.BlockNumber
+	}
+	ub.q = logs
+
+	prommetrics.AutomationLogsInLogBuffer.Add(float64(added))
+
+	return added, ub.clean(offsetBlock)
+}
+
+func (ub *upkeepLogBuffer) clean(offsetBlock int64) int {
+	maxLogs := int(ub.maxLogs.Load())
+
+	// sort.SliceStable(updated, func(i, j int) bool {
+	// 	return LogSorter(updated[i], updated[j])
+	// })
+	updated := make([]logpoller.Log, 0, maxLogs)
+	var dropped int
+	for _, l := range ub.q {
+		if l.BlockNumber > offsetBlock {
+			if len(updated) > maxLogs-1 {
+				prommetrics.AutomationLogsInLogBuffer.Dec()
+				// TODO: check if we should clean visited as well
+				ub.lggr.Debugw("Reached log buffer limits, dropping log", "blockNumber", l.BlockNumber,
+					"blockHash", l.BlockHash, "txHash", l.TxHash, "logIndex", l.LogIndex, "len updated", len(updated), "maxLogs", maxLogs)
+				dropped++
+			} else {
+				updated = append(updated, l)
+			}
+		} else {
+			prommetrics.AutomationLogsInLogBuffer.Dec()
+			// old logs are ignored and removed from visited
+			logid := logID(l)
+			delete(ub.visited, logid)
+		}
+	}
+	ub.q = updated
+
+	for lid, block := range ub.visited {
+		if block <= offsetBlock {
+			delete(ub.visited, lid)
+		}
+	}
+
+	return dropped
+}
+
+func logID(l logpoller.Log) string {
```

Comment:
We might have functionality that already does this; but also, does it make sense for this function to exist on the logpoller.Log type?

Reply:
This function uses `LogTriggerExtension.LogIdentifier()` to calculate the id.